File: gacyber_toolkit/CheatSheets/OS_and_Service_Version_Discovery.rb
#!/usr/bin/env ruby
require 'terminal-table'

def os_and_service_discovery(target_ip)
  puts "\n[+] Performing OS and Service Version Discovery on #{target_ip}...\n"

  # OS Discovery
  puts "[*] Detecting Operating System..."
  os_command = "nmap -O #{target_ip}" # Command to detect OS
  os_result = `#{os_command}`

  os_details = "Not Available"
  if os_result.include?("OS details")
    os_details = os_result.match(/OS details: (.+)/)[1] rescue "Not Available"
    puts "  OS Detected: #{os_details}"
  else
    puts "  No OS information detected."
  end

  # Service Version Discovery
  puts "\n[*] Detecting Service Versions on Open Ports..."
  service_command = "nmap -sV #{target_ip}" # Command to detect service versions
  service_result = `#{service_command}`

  # Parse the nmap output for open ports and services
  open_ports = service_result.scan(/^(\d+\/\w+)\s+open\s+([\w-]+)\s+(.+)/)

  if open_ports.any?
    # Prepare data for the table
    rows = open_ports.map do |port, service, version|
      [port, service, version.strip]
    end

    # Display the table
    table = Terminal::Table.new(
      title: "Service Version Information",
      headings: ['Port', 'Service', 'Version'],
      rows: rows
    )
    puts table
  else
    puts "  No services detected on open ports."
  end
end

# Main Entry Point
if ARGV.empty?
  puts "Usage: ruby os_service_discovery.rb <target_ip>"
  exit
end

target_ip = ARGV[0]
os_and_service_discovery(target_ip)

File: gacyber_toolkit/CheatSheets/OS_and_Service_Version_Discovery.py
import nmap
from prettytable import PrettyTable

def discover_os_and_services(target_ip):
    """Perform OS and service version discovery on the target host."""
    nm = nmap.PortScanner()

    print(f"\n[+] Scanning target: {target_ip} for OS and service discovery...\n")
    
    try:
        # OS Discovery
        print("[*] Detecting operating system...")
        nm.scan(target_ip, arguments='-O')  # -O flag for OS detection
        if 'osmatch' in nm[target_ip]:
            print("Operating System(s) Detected:")
            for os in nm[target_ip]['osmatch']:
                print(f"  - {os['name']} (Accuracy: {os['accuracy']}%)")
        else:
            print("  No OS information detected.")

        # Service Version Discovery
        print("\n[*] Detecting service versions on open ports...")
        nm.scan(target_ip, arguments='-sV')  # -sV flag for service version detection
        table = PrettyTable()
        table.field_names = ["Port", "Protocol", "Service", "Product", "Version"]
        if nm[target_ip].all_protocols():
            for protocol in nm[target_ip].all_protocols():
                ports = nm[target_ip][protocol].keys()
                for port in sorted(ports):
                    service = nm[target_ip][protocol][port]
                    table.add_row([
                        port,
                        protocol,
                        service['name'],
                        service.get('product', 'N/A'),
                        service.get('version', 'N/A')
                    ])
            print(table)
        else:
            print("  No services detected.")

    except Exception as e:
        print(f"[-] Error: {e}")

if __name__ == "__main__":
    target_ip = input("Enter the target IP address: ").strip()
    discover_os_and_services(target_ip)